presto-main/src/main/java/io/prestosql/sql/planner/optimizations/QueryCardinalityUtil.java

```diff
@@ -106,10 +106,33 @@ public CardinalityExtractorPlanVisitor(Lookup lookup)
         @Override
         public Range<Long> visitAggregation(AggregationNode node, Void context)
         {
-            if (node.hasEmptyGroupingSet()) {
+            if (node.hasEmptyGroupingSet() && node.getGroupingSetCount() == 1) {
+                // only single default aggregation which will produce exactly single row
                 return Range.singleton(1L);
             }
-            return Range.atLeast(0L);
+
+            Range<Long> sourceCardinalityRange = node.getSource().accept(this, null);
+
+            if (node.hasDefaultOutput() || sourceCardinalityRange.lowerEndpoint() > 0) {
```

Comment:
I find this code hard to read with all the cases and branches.

Maybe we can simply to something like:
```java
long lower;
if (node.hasDefaultOutput() || sourceCardinalityRange.lowerEndpoint() > 0) {
    lower = 1;
}
else {
    lower = 0;
}

if (sourceCardinalityRange.hasUpperBound()) {
    long upper = Math.max(lower, sourceCardinalityRange.upperEndpoint());
    return Range.closed(lower, upper);
}

return Range.atLeast(lower);
```